Add no-public-ip option to instance groups

moleksyuk opened the pull request:
This pull request contains changes related to issue #266

justinsb commented on upup/pkg/api/cluster.go:
```diff
@@ -103,6 +103,8 @@ type ClusterSpec struct {
 	//  * enable debugging handlers on the master, so kubectl logs works
 	IsolateMasters *bool `json:"isolateMasters,omitempty"`
 
+	AssociatePublicIP *bool `json:"associatePublicIp,omitempty"`
```
Wondering if this should be on the InstanceGroup spec instead?

moleksyuk replied:
Will move it to `upup/pkg/api/instancegroup.go`

justinsb commented on upup/pkg/fi/cloudup/apply_cluster.go:
```diff
@@ -230,7 +230,9 @@ func (c *ApplyClusterCmd) Run() error {
 				"autoscalingGroup":    &awstasks.AutoscalingGroup{},
-				"launchConfiguration": &awstasks.LaunchConfiguration{},
+				"launchConfiguration": &awstasks.LaunchConfiguration{
+					AssociatePublicIP: c.Cluster.Spec.AssociatePublicIP,
```
Ah ... this isn't how it was intended to be used.  Instead, you should use a template helper function, which you define here:

https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/cloudup/template_functions.go#L55

Then if it's a cluster spec that's pretty easy, because there's only one of them.  If it is on the InstanceGroup spec though, you can just pass the current instance group in as a variable like this example:

https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/_aws/master/_master_single/master_single.yaml#L41

justinsb commented on upup/pkg/fi/cloudup/template_functions.go:
```diff
@@ -77,6 +78,10 @@ func (tf *TemplateFunctions) AddTo(dest template.FuncMap) {
 	dest["IAMServiceEC2"] = tf.IAMServiceEC2
+
+	dest["AssociatePublicIP"] = func() bool {
+		return fi.BoolValue(tf.cluster.Spec.AssociatePublicIP)
```
Oh, yes, like this :-)

So you don't need to pass in AssociatePublicIP in apply_cluster :-)

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -354,6 +357,12 @@ func (c *CreateClusterCmd) Run(args []string) error {
 	}
 
+	if c.NoPublicIP {
+		fullCluster.Spec.AssociatePublicIP = fi.Bool(false);
```
I think it would be better if the flag was named similarly to the spec option; the flags are meant as shortcuts and I'd like to keep their names as similar as possible to help people discover this.  So you would pass `--associate-public-ip=false` (true would be the default).

WDYT?

moleksyuk replied:
My first implementation before pull request was in such way (I look at your `--yes` parameter which is bool) and thought it would be great to implement it in a similar way.
Also in general by default all bool variables are false, so have `--associate-public-ip` brings simple name but will break backward compatibility which is not good.
So for me `--no-public-ip` looks for me more pretty and understandable for enabling/disabling some options and looks identical like `--yes`.

So maybe it would be great to rename it to `--no-associate-public-ip` and by default, it will be `true` and backward compatible.

justinsb commented:
Looks really close, and nice job figuring out the various layers at play here!  Hopefully we can refactor this to make it simpler after the first stable release, and I'll probably be bothering you to ask for your feedback at that time.

I had a few thoughts which I put as inline comments, in particular that this should probably be a per-instancegroup setting?

On the one hand, I could see people mixing: having public ips for the masters only, or maybe for a set of nodes which run ingress.

On the other hand, the user must set it on each instancegroup, which is less convenient.

moleksyuk commented:
Hi @justinsb 
I've finished with all changes you mentioned above.
Could you please review it again.

And I renamed parameter to `--no-associate-public-ip` as I described in [this comment.](https://github.com/kubernetes/kops/pull/273#discussion_r73710099)

davekonopka commented:
Looking through the cloudup AWS models instances end up in a public subnet with `0.0.0.0` routed to an Internet Gateway. Disabling public ip's for instances with this routing setup would prevent outbound Internet access from nodes. Is there a way to override subnets/routing tables to place nodes on private subnets with NAT'ed routes?

justinsb commented on cmd/kops/create_cluster.go:
```diff
@@ -80,6 +81,8 @@ func init() {
 	cmd.Flags().StringVar(&createCluster.OutDir, "out", "", "Path to write any local output")
 	cmd.Flags().StringVar(&createCluster.AdminAccess, "admin-access", "", "Restrict access to admin endpoints (SSH, HTTPS) to this CIDR.  If not set, access will not be restricted by IP.")
+
+	cmd.Flags().BoolVar(&createCluster.NoAssociatePublicIP, "no-associate-public-ip", false, "Specify --no-associate-public-ip to disable association of public IP for master ASG and nodes.")
```
Just wondering why you went with "NoAssociatePublicIP" instead of "AssociatePublicIP".

If it was because you didn't want it to show up in `kops edit instancegroup`, I don't think it's a problem.

But maybe it was because you wanted it to default to true in the template?  That is interesting and might well require a helper function...

justinsb replied:
Actually, a better idea... what about adding a default value to `func (r *InstanceGroupRegistry) Find(name string) (*InstanceGroup, error) {` in registry.go.

The cleanest way might be to create a method on InstanceGroup called `SetDefaults` or similar, and then call it from Find in registry.go, and also in populate_instancegroup_spec.go.

We essentially want a field with a default value of true, but I don't know of an easier way.  And we do something similar in k8s.

moleksyuk commented:
@justinsb 
I've changed field from `--no-associate-public-ip` to `--associate-public-ip` and added default value in `populate_instancegroup_spec.go`.

Unfortunately, I didn't get your point about Find method in registry.go.
Why do we need to populate there with a default value? As I see right now (I assume you are talking about `Find` method in `upup/pkg/api/registry.go:188`) there is no any population with defaults for other values (for example a new one field called 'Image'). Should it be also populated with default from this method?

oviis commented:
Hi, do you know if/when this feature is coming to kops?
I think, if you want to build big clusters and also have some security, you cannot give any node an external IP. 
Also as @davekonopka asked, is there a planned way to route the traffic over nat eip's?

@moleksyuk @justinsb  Thanks for your great work guys, i really love it :-)

justinsb commented:
LGTM - thanks @moleksyuk :-)  As @oviis said, a great step towards isolated clusters!
